mlserver/batching/requests.py

```diff
@@ -0,0 +1,121 @@
+from operator import mul
+from collections import defaultdict
+from typing import Any, Dict, Iterable, List, Union
+from functools import reduce
+
+from ..types import (
+    InferenceRequest,
+    InferenceResponse,
+    RequestInput,
+    ResponseOutput,
+)
+
+
+def _get_data(payload: Union[RequestInput, ResponseOutput]):
+    return getattr(payload.data, "__root__", payload.data)
+
+
+def _infer_elem_size(shape: List[int]) -> int:
+    # TODO: Allow to use a different batch dimension
+    return reduce(mul, shape[1:], 1)
+
+
+def _merge_data(request_inputs: List[RequestInput]) -> Any:
+    all_data = [_get_data(request_input) for request_input in request_inputs]
+
+    sampled_datum = all_data[0]
+
+    if isinstance(sampled_datum, str):
+        return "".join(all_data)
+
+    if isinstance(sampled_datum, bytes):
+        return b"".join(all_data)
+
+    if isinstance(sampled_datum, list):
+        return sum(all_data, [])
+
+    # TODO: Should we raise an error if we couldn't merge the data?
+    return all_data
+
+
+def _split_data(response_output: ResponseOutput) -> Any:
+    element_size = _infer_elem_size(response_output.shape)
+    merged_data = _get_data(response_output)
+
+    # TODO: Don't rely on array to have been flattened
+    for i in range(0, len(merged_data), element_size):
+        yield merged_data[i : i + element_size]
+
+
+class BatchedRequests:
+    def __init__(self, inference_requests: List[InferenceRequest] = []):
+        self._inference_requests = inference_requests
+        self.merged_request = self._merge_requests()
+        self._prediction_ids: List[str] = [  # type: ignore
+            req.id for req in self._inference_requests  # type: ignore
+        ]
+
+    def _merge_requests(self) -> InferenceRequest:
+        inputs_index: Dict[str, List[RequestInput]] = defaultdict(list)
+
+        for inference_request in self._inference_requests:
+            # TODO: What should happen if UID is empty?
```

Comment:
I guess it should always be populated, as it's either provided or should be automaticallyy generated (that's what happens in executor)

Reply:
Yuuup, good point. It should be safe to assume that it's going to be set up. In fact, I think that the layer above this method actually sets it if it's unset.

Reply:
I guess perhaps the issue would be more if the user sends user defined IDs that happen to not be unique... I guess the issue here would be the assumption that the IDs are unique... perhaps here the comment to add is whether we'd want to add a separate internal unique identifier to avoid clashes?

Reply:
Right, it should now be changed to use an internal set of IDs (i.e. without relying on the external request IDs).